### PR TITLE
Feature/send IDS multipart messages

### LINKED ISSUE #233: RemoteMessageDispatcher and MessageSender for retrieving IDS descriptions
For the purpose of enabling communication of one connector to another using the IDS multipart protocol another [RemoteMessageDispatcher](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/spi/src/main/java/org/eclipse/dataspaceconnector/spi/message/RemoteMessageDispatcher.java) implementation is needed. Furthermore this component needs to be enabled to issue IDS [DescriptionRequestMessage](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#DescriptionRequestMessage)s for retrieving (self)descriptions of

* the [Connector](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#Connector)
* the [Catalog](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#Catalog)
* an Asset's [Resource](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#Resource)s
* an Asset's [Artifact](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#Artifact)s
* an Asset's [Representation](https://international-data-spaces-association.github.io/InformationModel/docs/index.html#Representation)s

Anologuous to the [ids-api-multipart-endpoint-v1](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/main/data-protocols/ids/ids-api-multipart-endpoint-v1) another gradle module underneath the [data-protocols/ids ](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/main/data-protocols/ids) tree must be created. This module then can utilize the [object transformation infrastructure](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/tree/main/data-protocols/ids/ids-transform-v1) to map between IDS objects and EDC domain objects.

This issue is related but not dependent to #234.

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

### PR DESCRIPTION
**Pull request by ronjaquensel**
Closes #233 

This PR enables the EDC to send IDS multipart messages. It adds a new module `ids-api-multipart-dispatcher-v1`, which contains IdsMessageSender implementations as well as a RemoteMessageDispatcher implementation for IDS multipart.  Senders have been implemented for the following message types:

* DescriptionRequestMessage
* ArtifactRequestMessage
* ContractRequestMessage
* ContractAgreementMessage

### CODE REVIEW

File: data-protocols/ids/ids-api-multipart-dispatcher-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/sender/MultipartArtifactRequestSender.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    protected Message buildMessageHeader(DataRequest request, DynamicAttributeToken token) throws Exception {
+        var asset = request.getAsset();
+        IdsId id = IdsId.Builder.newInstance()
+                .value(asset.getId())
+                .type(IdsType.ARTIFACT)
+                .build();
+        var transformationResult = transformerRegistry.transform(id, URI.class);
+        if (transformationResult.hasProblems()) {
+            throw new EdcException("Failed to create artifact ID from asset.");
```
Comment (jimmarino) [inline data-protocols/ids/ids-api-multipart-dispatcher-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/sender/MultipartArtifactRequestSender.java:73]: Not to change now (can be done in a later PR), but can we report the errors and return an error message to the client?

File: data-protocols/ids/ids-api-multipart-dispatcher-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/message/MultipartRequestInProcessResponse.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import com.fasterxml.jackson.annotation.JsonCreator;
+import de.fraunhofer.iais.eis.Message;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+
+//TODO define return type
+public class MultipartRequestInProcessResponse implements MultipartResponse<String> {
+
+    Message header;
```
Comment (jimmarino) [inline data-protocols/ids/ids-api-multipart-dispatcher-v1/src/main/java/org/eclipse/dataspaceconnector/ids/api/multipart/client/message/MultipartRequestInProcessResponse.java:26]: Not to change now (can be done in a later PR), but can this be private?

**Review by jimmarino — APPROVED** [review]
Awesome stuff @ronjaquensel! Can we merge this in now after @juliapampus reviews? I had two very minor comments, we can look at later since they are just small nits.

**Comment by denisneuling** [comment]
@jimmarino Ronja and me we had a chat and agreed, that I'll provide some minor refactoring. I was already about to push the changes, but underway transmit got denied. Guess we need to wait till the next morning.

<sub> Denis Neuling <denis.neuling@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md)

**Comment by jimmarino** [comment]
@denisneuling is it essential refactoring or can we merge this first? If it's not critical, I think it would be good to merge as-is assuming @juliapampus is OK with her review.

**Comment by denisneuling** [comment]
@jimmarino There is a bug with the multipart-controller the dispatcher is talking to (expecting RequestMessages only, but we're sending ResponseMessages too now). Also there is a flaw how the requesting connector's id is set. 
So getting the refactoring in is essential for it to function properly. No rush, it's all prepared, just need to get the invitation to Ronja's repo tomorrow morning, will push then and immediately ping you afterwards. 
I'll then prepare and reach in the control API for initiating the data transfer then.

**Comment by jimmarino** [comment]
> @jimmarino There is a bug with the MultipartController the dispatcher is talking to (expecting RequestMessages only, but we're sending ResponseMessages too now). Also there is a flaw how the requesting connector's id is set. So getting the refactoring in is essential for it to function properly. No pressure, it's all prepared, just need to get the invitation to Ronja's repo tomorrow morning, will push then and immediately ping you afterwards. I'll then prepare and reach in the control API for initiating the data transfer then.

OK great, thx

**Comment by juliapampus** [comment]
Ich have some questions:
- Why are the `DataRequestMessageSender` and `QueryMessageSender` in the `ids-core` and the other ones in other ids packages? Are they anyway needed anymore?
- Why is `DescriptionresponseMessageUtil` in another package than the `RejectionMessageUtil`? The package structure is not that clear, at least for me.
- Why is `MultipartRequest` in the endpoint package? What is the idea to have two packages for the multipart message sending and handling when they are not clearly separated and both contain sending and processing classes?

**Comment by DominikPinsel** [comment]
> > @jimmarino There is a bug with the MultipartController the dispatcher is talking to (expecting RequestMessages only, but we're sending ResponseMessages too now). Also there is a flaw how the requesting connector's id is set. So getting the refactoring in is essential for it to function properly. No pressure, it's all prepared, just need to get the invitation to Ronja's repo tomorrow morning, will push then and immediately ping you afterwards. I'll then prepare and reach in the control API for initiating the data transfer then.
> 
> OK great, thx

I fixed this problem in my PR. Therefore we won't need to resolve it here.

**Comment by ronjaquensel** [comment]
Hi @juliapampus,

the sender classes in `ids-core` are from April/May. I suppose they were example implementations before there were other designated IDS modules. I'm not sure if they're used in some sample, therefore I just left them as is. They should probably be re-implemented once we have a module for IDS Rest.

**Comment by DominikPinsel** [comment]
> Ich have some questions:
> 
> * Why are the `DataRequestMessageSender` and `QueryMessageSender` in the `ids-core` and the other ones in other ids packages? Are they anyway needed anymore?
> * Why is `DescriptionresponseMessageUtil` in another package than the `RejectionMessageUtil`? The package structure is not that clear, at least for me.
> * Why is `MultipartRequest` in the endpoint package? What is the idea to have two packages for the multipart message sending and handling when they are not clearly separated and both contain sending and processing classes?

I will get rid of the DataRequestMessageSender in my PR.

**Comment by DominikPinsel** [comment]
> Ich have some questions:
> 
> * Why are the `DataRequestMessageSender` and `QueryMessageSender` in the `ids-core` and the other ones in other ids packages? Are they anyway needed anymore?
> * Why is `DescriptionresponseMessageUtil` in another package than the `RejectionMessageUtil`? The package structure is not that clear, at least for me.
> * Why is `MultipartRequest` in the endpoint package? What is the idea to have two packages for the multipart message sending and handling when they are not clearly separated and both contain sending and processing classes?

I have an alternative implementation of the `DataRequestMessageSender` in my PR. Therefore we should be able to remove this after my PR is closed.

**Comment by juliapampus** [comment]
In addition, comments are missing. I know this is annoying, but in the other modules, at least the classes and most important methods have a description that shortly hint on what the idea was. This maybe also helps to clean up the module structure and recognize inconsistencies.

**Comment by denisneuling** [comment]
> * Why are the `DataRequestMessageSender` and `QueryMessageSender` in the `ids-core` and the other ones in other ids packages? Are they anyway needed anymore?

This is some old implementation and is still used within demo applications. This will be removed in a subsequent PR. 

> * Why is `DescriptionresponseMessageUtil` in another package than the `RejectionMessageUtil`? The package structure is not that clear, at least for me.

There was lots of discussing going on how to organize the components, finally we found consent doing it this way (, I am btw not a big fan of). Anyway changing this within this PR would be out of scope of #275, don't you agree?  

> * Why is `MultipartRequest` in the endpoint package? What is the idea to have two packages for the multipart message sending and handling when they are not clearly separated and both contain sending and processing classes?

The `MultipartRequest` and `MultipartResponse` objects are wrapper objects used in the multipart endpoint package. Their naming was a natural choice to reflect theiy're wrapping request/response data. We can rename them / reorganize the class structure to make them package visible only in another PR if you like.

**Comment by juliapampus** [comment]
> > * Why are the `DataRequestMessageSender` and `QueryMessageSender` in the `ids-core` and the other ones in other ids packages? Are they anyway needed anymore?
> 
> This is some old implementation and is still used within demo applications. This will be removed in a subsequent PR.
> 
> > * Why is `DescriptionresponseMessageUtil` in another package than the `RejectionMessageUtil`? The package structure is not that clear, at least for me.
> 
> There was lots of discussing going on how to organize the components, finally we found consent doing it this way (, I am btw also not a big fan of). Anyway changing this within this PR would be out of scope of #275, don't you agree?
> 
> > * Why is `MultipartRequest` in the endpoint package? What is the idea to have two packages for the multipart message sending and handling when they are not clearly separated and both contain sending and processing classes?
> 
> The `MultipartRequest` and `MultipartResponse` objects are wrapper objects used in the multipart endpoint package. Their naming was a natural choice to reflect theiy're wrapping request/response data. We can rename them / reorganize the class structure to make them package visible only in another PR if you like.

I agree, we should do a restructuring in another PR. But I think we definetely need to clean up this ids module... @ronjaquensel is currently about to add some documentation. We can merge this afterwards.

**Comment by ronjaquensel** [comment]
I've added Javadoc on class level as well as for abstract methods.

**Review by juliapampus — APPROVED** [review]

**Pull request merged**